scipy/linalg/_solvers.py

```diff
@@ -302,71 +313,162 @@ def solve_discrete_are(a, b, q, r):
-    u21 = u[m//2:m, 0:n//2]
-    u11i = inv(u11)
-
-    return np.dot(u21, u11i)
+    
+    a = np.atleast_2d(_asarray_validated(a))
+    b = np.atleast_2d(_asarray_validated(b))
+    q = np.atleast_2d(_asarray_validated(q))
+    r = np.atleast_2d(_asarray_validated(r))
+    
+    # Get the correct data types otherwise Numpy complains about pushing 
+    # complex numbers into real arrays.
+    r_or_c = complex if np.iscomplexobj(b) else float
+
+    for ind, x in enumerate((a,q,r)):
+        if np.iscomplexobj(x):
+            r_or_c = complex
+            
+        if not np.equal(*x.shape):
+            raise ValueError("Matrix {} should be square.".format("aqr"[ind]))
+    
+    # Shape consistency checks
+    m, n = b.shape
+
+    if m != a.shape[0]:
+        raise ValueError("Matrix a and b should have the same number of rows.")
+        
+    if m != q.shape[0]:
+        raise ValueError("Matrix a and q should have the same shape.")
+            
+    if n != r.shape[0]:
+        raise ValueError("Matrix b and r should have the same number of cols.")
+    
+    # Check if the data is (sufficiently) hermitian
+    if norm(q - q.conj().T,1) > np.spacing(norm(q,1))*100:
+        raise ValueError("Matrix q should be symmetric/hermitian.")
+
+    if norm(r - r.conj().T,1) > np.spacing(norm(r,1))*100:
+        raise ValueError("Matrix r should be symmetric/hermitian.")
+
+    # Form the matrix pencil        
+    H = np.zeros((2*m + n, 2*m + n), dtype = r_or_c)
+    H[:m, :m] = a
+    H[:m, 2*m:] = b
+    H[m:2*m, :m] = -q
+    H[m:2*m, m:2*m] = np.eye(m)
+    H[2*m:, 2*m:] = r
+
+    J = np.zeros_like(H, dtype = r_or_c)
+    J[:m, :m] = np.eye(m)
+    J[m:2*m, m:2*m] = a.conj().T
+    J[2*m:, m:2*m] = -b.conj().T
+
+    # Deflate the pencil by the R column ala Ref. [1]
+    q_of_qr, _ = qr(H[:,-n:])
+    H = q_of_qr[:,n:].conj().T.dot(H[:,:2*m])
+    J = q_of_qr[:,n:].conj().T.dot(J[:,:2*m])
+    
+    # Decide on which output type is needed for QZ 
+    out_str = 'real' if r_or_c == float else 'complex'
+    
+    _, _, _, _, _, u = ordqz(H, J, sort = 'iuc', overwrite_a=True, 
+                    overwrite_b=True, check_finite=False, output=out_str)
+    
+    # Get the relevant parts of the stable subspace basis
+    u00 = u[:m, :m]
+    u10 = u[m:2*m, :m]
+    
+    # Solve via back-substituion after checking the condition of u00 
+    up, ul, uu = lu(u00)
+
+    if 1/cond(uu) < np.spacing(1.):
+        raise LinAlgError('Failed to find a finite solution.')
+    
+    # Exploit the triangular structure
+    x = solve_triangular(ul.conj().T,
+                         solve_triangular(uu.conj().T, 
+                                          u10.conj().T,
+                                          lower=True),
+                         unit_diagonal=True,
+                         ).conj().T.dot(up.conj().T)
+
+    # Check the deviation from symmetry for success
+    u_sym = u00.conj().T.dot(u10)
+    u_sym -= u_sym.conj().T
```

Comment:
It's the issue about Numpy arrays and views to numpy arrays, eg `b = a[::2]; b[0] = 1` modifies also `a`. I think it's mentioned in most numpy tutorials. With in-place operations etc., you want to make sure the output operand is not a view of any kind of the input.

However, I think this caveat will go away in near future https://github.com/numpy/numpy/pull/8043. There are also long threads about this in the numpy issue tracker, but I'd recommend those only for entertainment purposes now that there's a consensus on how the issue will be addressed.